### PR TITLE
add more metrics for slow commit log diagnostics

### PR DESCRIPTION
**Pull request by tonyxuqqi**
### What is changed and how it works?

Issue Number: Ref #15175

What's Changed:

```commit-message
Add more metrics for slow commit log duration investigation. 
In this PR, it adds raft message process wait duration and exposes raft
message recv by store counter. Together with raft-engine write duration,
we can further narrow reason of the commit log duration. With this PR,
we still cannot tell if the slowness comes from network or
raft-client's (grpc client).  
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Release note 

```release-note
None
```

### CODE REVIEW

File: metrics/grafana/tikv_details.json [hunk 1]
```diff
@@ -5908,6 +5916,111 @@
...
           }
+        },
+        {
+          "aliasColors": {},
+          "bars": false,
+          "dashLength": 10,
+          "dashes": false,
+          "datasource": "${DS_TEST-CLUSTER}",
+          "decimals": 1,
+          "description": "The count of gRPC raft message",
```
Comment (v01dstar) [inline metrics/grafana/tikv_details.json:5927]: Is this change intended?

Comment (tonyxuqqi) [reply metrics/grafana/tikv_details.json:5927]: yes, added a new chart under grpc panel

**Review by v01dstar — APPROVED** [review]
LGTM

File: components/raftstore/src/store/local_metrics.rs [hunk 2]
```diff
@@ -113,6 +113,7 @@ pub struct RaftMetrics {
     // local histogram
     pub store_time: LocalHistogram,
     pub propose_wait_time: LocalHistogram,
+    pub process_wait_time: LocalHistogram,
```
Comment (bufferflies) [inline components/raftstore/src/store/local_metrics.rs:116]: how about rename commit_rpc_time or add comment to detail the duration.

File: components/raftstore/src/store/metrics.rs [hunk 3]
```diff
@@ -551,6 +551,13 @@ lazy_static! {
             exponential_buckets(0.00001, 2.0, 26).unwrap()
         ).unwrap();
 
+    pub static ref RAFT_MESSAGE_WAIT_TIME_HISTOGRAM: Histogram =
```
Comment (bufferflies) [inline components/raftstore/src/store/metrics.rs:554]: Does it only include send message in rpc not include any handler?

**Review by bufferflies — APPROVED** [review]

File: components/raftstore-v2/src/router/message.rs [hunk 4]
```diff
@@ -157,7 +157,7 @@ pub enum PeerMsg {
     /// Raft message is the message sent between raft nodes in the same
     /// raft group. Messages need to be redirected to raftstore if target
     /// peer doesn't exist.
-    RaftMessage(Box<RaftMessage>),
+    RaftMessage(Box<RaftMessage>, Option<Instant>),
```
Comment (LykxSassinator) [inline components/raftstore-v2/src/router/message.rs:160]: How about use a object to encapsulate the metrics?
Such as:
```Rust
struct Metrics {
    send_time: Instant,
    /// ... add more timestamp.
}
```
Or just like the implementation in `raftstore-v1` -- `InspectRaftMessage` does.
https://github.com/tikv/tikv/blob/b4e0bf7bab6ad395b74c0be938119d82ded4cd2a/components/raftstore/src/store/msg.rs#L733-L736
Add one extra `send_time: Instant` into it?

File: components/raftstore/src/store/msg.rs [hunk 5]
```diff
@@ -741,7 +741,7 @@ pub enum PeerMsg<EK: KvEngine> {
     /// Raft message is the message sent between raft nodes in the same
     /// raft group. Messages need to be redirected to raftstore if target
     /// peer doesn't exist.
-    RaftMessage(InspectedRaftMessage),
+    RaftMessage(InspectedRaftMessage, Option<Instant>),
```
Comment (LykxSassinator) [inline components/raftstore/src/store/msg.rs:744]: Ditto. How about add one extra metering metric `send_time` into `InspectedRaftMessage`.

Comment (tonyxuqqi) [reply components/raftstore/src/store/msg.rs:744]: I think the send_time is only interested when InspectedRaftMessage is inside the Peer::RaftMessage.

File: components/raftstore-v2/src/router/message.rs (hunk 4 above)
Comment (tonyxuqqi) [reply components/raftstore-v2/src/router/message.rs:160]: See the comments below.
